feat(parsers/csv): Add metadata support to CSV parser plugin

etycomputer opened the pull request:
### Required for all PRs:

<!-- Complete the tasks in the following list. Change [ ] to [x] to
show completion. -->

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

<!-- Link to issues that describe the need for the change. Issues
should include context that will help reviewers understand why the
change is needed.

Make sure to link issues and using a keyword like "resolves #1234".
https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->

resolves #10079

<!-- Finally, include a summary of the code change itself. This
description should tell reviewers how the issues were resolved.

example: Fixed an off by one error in counter variable in type FooBar.

example: Added an input plugin to gather yak shaving metrics using
golang library yaktech/shaver. -->

Added support to parse metadata for the CSV parser plugin.

etycomputer commented:
Hi @sspaink ,

I have another Linter issue. It is erroring things that I have not changed in the READEME.md file.

Would you be able to have a look, please?

I know how to fix the issue. But I am not sure if I should.

Thanks,
Ehsan

sspaink commented on plugins/parsers/csv/README.md:
```diff
@@ -33,8 +33,21 @@ values.
   csv_skip_rows = 0
+  
+  
+  ## Indicates the number of rows that parse for metadata before looking for header information.
+  ## Indicates how many rows to parse for metadata. By default, the parser assumes there are no 
+  ## metadata rows to parse. If set, the parser would use the regex provided in the csv_metadata_regex 
+  ## to look for metadata.
```
```suggestion
  ## Indicates the number of rows to parse as metadata before looking for header information. 
  ## By default, the parser assumes there are no metadata rows to parse. 
  ## If set, the parser would use the regex provided in the csv_metadata_regex  to look for metadata.
```

sspaink commented on plugins/parsers/csv/README.md:
```diff
@@ -33,8 +33,21 @@ values.
+  ## metadata rows to parse. If set, the parser would use the regex provided in the csv_metadata_regex 
+  ## to look for metadata.
+  # csv_metadata_rows = 0
+  ## A list of regex patters with two subgroups named key and value. If csv_metadata_rows is set,
+  ## csv_metadata_regex must contain atleaset one regex patters. Metadata key and value must be
+  ## on the same row.
+  ## For example:
+  ## Useing "(?P<key>\w+)[=](?P<value>\w+)" to parse "key1=value1" would return key1 and value1
```
```suggestion
  ## A list of regex patterns with two subgroups named key and value. If csv_metadata_rows is set,
  ## csv_metadata_regex must contain at least one regex patterns. Metadata key and value must be
  ## on the same row.
  ## For example:
  ## Using "(?P<key>\w+)[=](?P<value>\w+)" to parse "key1=value1" would return key1 and value1
```

sspaink reviewed (changes requested):
@etycomputer thanks for working on this new feature, noticed some minor typos in the `README.md` so I've got some requested changed. To answer your questions:

- The bot sometimes get the labels wrong, its a todo item to get it more accurate by using the conventional commit message. I've fixed the labels manually in the meantime :)

- The markdown linter unfortunately doesn't just check your changes but the whole file, I decided to fix the linter errors in a separate pull request: https://github.com/influxdata/telegraf/pull/10093

etycomputer commented:
Hi @sspaink,
Thanks for fixing the markdown linter issue on #10093 and your feedback regarding my new feature.
I have finished working on this job. I have tried to add all the scenarios that I can think of in the unit tests.

The only part that I don't like is where I am using the different Readers. 
I have started learning to program in Go not more than three weeks, so if there is a more efficient way to reach the same goal please let me know.

Basically, the part that gets messy is when I am trying to read lines to skip or parse metadata. in these cases, I don't want to use the CSV reader and just want to use a basic reader.

I look forward to hearing from you,

Kind regards,
Ehsan

etycomputer commented:
Hi @sspaink,

When you have a few minutes could you have a look at this PR.
I have fully tested it with the CSV files that I needed to parse and it is now working very well.

I look forward to receiving your feedbacks.

Kind regards,
Ehsan

Hipska commented on plugins/parsers/csv/parser.go:
```diff
@@ -16,34 +19,111 @@ import (
-	DefaultTags map[string]string
+	TimeFunc      func() time.Time
+	DefaultTags   map[string]string
+	DefaultFields map[string]string
```
The map contains the actual metadata, `DefaultFields` confused me.
```suggestion
	Metadata map[string]string
```

etycomputer replied:
Just like default tags I had to keep the default fields that we have collected from the metadata section.

Hipska commented on plugins/parsers/csv/parser.go:
```diff
@@ -262,6 +383,11 @@ outer:
 		tags[k] = v
 	}
 
+	// add default fields
+	for k, v := range p.DefaultFields {
+		recordFields[k] = v
+	}
```
So if a field with the same name would already exist, then you just override it with data from the file's metadata?

etycomputer replied:
Good point @Hipska. But this is the same as tags too. If you have global tags and the same column tags exist, the global tags would overwrite the column tags.


I could move both to the top but that would be a big change. 

I think this should be addressed on a different task.

Hipska replied:
You are right, that's indeed how it already works for global tags. But here we are talking about fields, and we can keep the current implementation, but I would log a warning (or lower severity) message stating that the field has been overwritten. So the user knows how it comes the field value seems to be the same for every row in the csv.

etycomputer replied:
Very good point and idea.

Hipska replied:
Could it be you didn't push this change yet?

Hipska commented on plugins/parsers/csv/README.md:
```diff
@@ -33,8 +33,35 @@ values.
+  ## to parse "# Version = 1.1" "Creation Date: '2021-11-01'"
+  ## the returned (key,value) pares would be as follows:
+  ## (Version,1.1), (Creation Date,2021-11-01)
+  cvs_metadata_trim_set = ""
```
`separators` is an array and `trim_set` is a string, that doesn't seem consistent..

etycomputer replied:
When trimming you give a string of characters, but when splitting key value pairs you need to have a unique set of separator strings ordered by their length.

Hipska replied:
@srebhan what are your thoughts on this?

srebhan replied:
I tend to agree with @etycomputer as this is a set of characters. Maybe we should name the option `csv_metadata_trim_characters`?

Hipska commented on plugins/parsers/csv/README.md:
```diff
@@ -33,8 +33,35 @@ values.
+  ## to parse "Key1,Value1" "Key2:Value2" "Key3=Value3"
+  ## the returned (key,value) pares would be as follows:
+  ## (Key1,Value1), (Key2,Value2), (Key3,Value3)
+  cvs_metadata_separators = [","]
```
Would `=` or `:` not be a better default value? It seems to me those are more commonly used to express key and values.

etycomputer replied:
The reason I have used ',' is because its a CSV file and logically you would separate the key value with commas.

For what I need this feature, I need ':' and '='. But if I had the option to rewrite the code I would not use two different separator characters.

Hipska replied:
In CSV, you separate the columns/values with `,` There are no key-value pairs. I would still use either `=` or `:` as default example as that will be more clear to users what it does.

Hipska commented on plugins/parsers/csv/README.md:
```diff
@@ -33,8 +33,35 @@ values.
+  ## For example:
+  ## cvs_metadata_separators = [",", ":", "="]
+  ## to parse "Key1,Value1" "Key2:Value2" "Key3=Value3"
+  ## the returned (key,value) pares would be as follows:
+  ## (Key1,Value1), (Key2,Value2), (Key3,Value3)
```
This section confuses me a bit. I think you better write the resulting tags/fields in influx format.

etycomputer replied:
Good idea. I will do that tomorrow.

Hipska replied:
You also might want to add here (or in the section for `metadata_rows`) that the key value pairs will be added as field by default and they can use `tag_columns` to convert them to tags.

etycomputer replied:
Done

Hipska reviewed (changes requested):
General idea looks good. Only it looks more logical that metadata would end up in metric _tags_ instead of metric _fields_. Or maybe have it as an option if you don't agree.

etycomputer commented:
> General idea looks good. Only it looks more logical that metadata would end up in metric _tags_ instead of metric _fields_. Or maybe have it as an option if you don't agree.

Hey @Hipska, thanks for your rapid feedback. I really appreciate it. Yes, I agree with you too. Metadata is commonly going to be saved a tags. But just in case this is not the case I have allowed the option for it to be a field by default and adding them as tags is possible by listing the tag keys.

The other reason behind considering the default for metadata to be fields is, we usually exactly know keys that are going to be tags but dynamic fields are harder to predict.

Finally, the current CSV parse considers all columns to be fields unless they are flagged as tags, I am following the same pattern.

Kind regards,
Ehsan

srebhan commented on plugins/parsers/csv/parser.go:
```diff
@@ -16,34 +19,113 @@ import (
 
 type TimeFunc func() time.Time
 
+type Metadata struct {
+	key   string
+	value string
+}
```
I think you should just use a `map[string]string` for this.

srebhan commented on plugins/parsers/csv/parser.go:
```diff
@@ -117,15 +214,41 @@ func (p *Parser) ParseLine(line string) (telegraf.Metric, error) {
+		}
+		p.MetadataRows--
+		m := p.ParseMetadataRow(line)
+		if m != nil {
+			for _, tagName := range p.TagColumns {
+				if tagName == m.key {
+					if p.DefaultTags == nil {
+						p.DefaultTags = make(map[string]string)
+					}
+					p.DefaultTags[m.key] = m.value
+					continue outer
+				}
+			}
+			if p.DefaultFields == nil {
+				p.DefaultFields = make(map[string]string)
+			}
+			p.DefaultFields[m.key] = m.value
+		}
```
TBH I don't get the meaning of this. Are you trying to decide of the found meta-data is a tag or field? If so, please don't. You can still convert the between them afterwards, using a processor. Please also avoid messing with `DefaultTags` as those are set by telegraf and you should not make any assumption whatsoever on when it is set.

srebhan commented on plugins/parsers/csv/parser.go:
```diff
@@ -259,9 +382,22 @@ outer:
 
 	// add default tags
 	for k, v := range p.DefaultTags {
+		_, exists := tags[k]
+		if exists {
+			p.Log.Warnf("Tag %s has been overwritten.", k)
+		}
```
What is this good for? That's the meaning of a _default_ IMO to be overwritten if you have other evidence. Please remove it as it has the potential to spam log-files for large number of similarly structured metrics.

Hipska replied:
It comes because he is overwriting the value with a 'default'. So other way round and informing user about that is a good thing IMHO.

srebhan commented on plugins/parsers/csv/parser.go:
```diff
@@ -259,9 +382,22 @@ outer:
 		tags[k] = v
 	}
 
+	// add default fields
+	for k, v := range p.DefaultFields {
+		_, exists := recordFields[k]
+		if exists {
+			p.Log.Warnf("Field %s has been overwritten.", k)
+		}
+		recordFields[k] = v
+	}
+
```
This should be removed completely IMO as we should _only_ add metadata as tags.

srebhan reviewed (commented):
Hey @etycomputer we are on the right way. However, I have some comments in the code.

srebhan commented on plugins/parsers/csv/parser.go:
```diff
@@ -17,33 +19,100 @@ import (
+func (record MetadataPattern) Swap(i, j int) {
+	record[i], record[j] = record[j], record[i]
+}
+func (record MetadataPattern) Less(i, j int) bool {
+	//if len(record[i]) == len(record[j]) {
+	//	return record[i] < record[j]
+	//}
+	return len(record[i]) > len(record[j])
+}
```
Hmm `Less` should return `true` if `i` is "less" than `j`. In this context it should return `true` if the record in `i` is _shorter_ than the one in `j`. I guess you are doing it this way as you need the _reverse_ sorting by length right? Anyway, a comment would help for others to understand this.

Furthermore, please delete the commented, dead code.

srebhan commented on plugins/parsers/csv/parser.go:
```diff
@@ -257,8 +354,21 @@ outer:
 		}
 	}
 
+	// add metadata fields
+	for k, v := range p.MetadataTags {
+		_, exists := tags[k]
+		if exists {
+			continue
+		}
+		tags[k] = v
+	}
+
 	// add default tags
 	for k, v := range p.DefaultTags {
+		_, exists := tags[k]
+		if exists {
+			continue
+		}
```
As discussed, we need to clarify if we really want to maintain this order of priorities...

srebhan replied:
Keep as is (default overwrites row values) and handle the order in a separate PR as commented below.

srebhan reviewed (commented):
Nice update @etycomputer. I want to ask you for one more comment (in the code) for the sorting to make sure future readers don't need to stare on the code asking themselves why less is more in this case... ;-P

etycomputer commented:
> General idea looks good. Only it looks more logical that metadata would end up in metric _tags_ instead of metric _fields_. Or maybe have it as an option if you don't agree.

Hi @Hipska,
Metadata are now saved as tags and users can convert them to fields if they need to by using a processor.

Hipska commented on plugins/parsers/csv/README.md:
````diff
@@ -128,4 +146,34 @@ Output:
+
+Input:
+
+```shell
````
```suggestion
```text
```
maybe `csv` is also possible?

Hipska commented on plugins/parsers/csv/README.md:
````diff
@@ -128,4 +146,34 @@ Output:
+
+Output:
+
+```shell
````
```suggestion
```text
```

srebhan commented:
Hey @etycomputer, we had a chance to chat with the core-devs. The punchline is to merge this PR with the "wrong" tag-overwriting priorities and then create another PR adding an option to choose the priorities (e.g. `rows_overwrite_default_tags = true/false`) with the current behavior being the default. Are you interested in contributing the mentioned PR?

etycomputer commented:
Hi @srebhan,
Thanks for following up my PR.
That is a very good idea.
I am on holiday for a few days but as soon as I get a few minutes I will make the changes.
I will also create the new PR for adding new switch.

etycomputer commented:
I have created a new FR https://github.com/influxdata/telegraf/issues/10219

etycomputer commented:
I have removed the priority feature from this PR. Hopefully, this feature branch will now be ready to merge into the master branch.

Hipska commented:
I still have some unresolved comments

etycomputer commented:
> I still have some unresolved comments

Sorry about that I must have missed those. They should be fixed now.

srebhan commented on plugins/parsers/csv/README.md:
```diff
@@ -33,8 +33,26 @@ values.
+  ## A list of metadata separators. If csv_metadata_rows is set,
+  ## cvs_metadata_separators must contain at least one separator.
+  ## Please note that separators are case sensitive and the sequence of the seperators are respected.
+  cvs_metadata_separators = [":", "="]
```
```suggestion
  csv_metadata_separators = [":", "="]
```

srebhan commented on plugins/parsers/csv/parser.go:
```diff
@@ -17,33 +19,98 @@ import (
+	TrimSpace          bool     `toml:"csv_trim_space"`
+	SkipValues         []string `toml:"csv_skip_values"`
+	MetadataRows       int      `toml:"csv_metadata_rows"`
+	MetadataSeparators []string `toml:"cvs_metadata_separators"`
+	MetadataTrimSet    string   `toml:"cvs_metadata_trim_set"`
```
```suggestion
	MetadataSeparators []string `toml:"csv_metadata_separators"`
	MetadataTrimSet    string   `toml:"csv_metadata_trim_set"`
```

srebhan commented on plugins/parsers/csv/parser.go:
```diff
@@ -17,33 +19,98 @@ import (
+	MetadataSeparators []string `toml:"cvs_metadata_separators"`
+	MetadataTrimSet    string   `toml:"cvs_metadata_trim_set"`
+
+	MetadataSeparatorList MetadataPattern
```
I think we should not export this variable.

srebhan commented on plugins/parsers/csv/parser.go:
```diff
@@ -17,33 +19,98 @@ import (
-	DefaultTags map[string]string
+	TimeFunc     func() time.Time
+	DefaultTags  map[string]string
+	MetadataTags map[string]string
```
Do we need to export those?

srebhan commented on plugins/parsers/csv/parser.go:
```diff
@@ -17,33 +19,98 @@ import (
 	*Config
 }
 
+type MetadataPattern []string
```
Do we need to export this one?

srebhan commented on plugins/parsers/csv/parser.go:
```diff
@@ -17,33 +19,98 @@ import (
+	return len(record[i]) > len(record[j])
+}
+
+func getUniqueMetadataPatterns(patternMap []string) MetadataPattern {
+	result := MetadataPattern{}
+	patternList := map[string]bool{}
+	for _, p := range patternMap {
+		if patternList[p] {
+			// Ignore further, duplicated entries
+			continue
+		}
+		patternList[p] = true
+		result = append(result, p)
+	}
+	return result
+}
+
+func (c *Config) initializeMetadataSeparators() error {
+	// initialize metadata
+	c.MetadataTags = map[string]string{}
+	c.MetadataSeparatorList = []string{}
+
+	if c.MetadataRows > 0 {
+		if len(c.MetadataSeparators) == 0 {
+			return fmt.Errorf("when csv_metadata_rows is defined, " +
+				"cvs_metadata_separators must have at least one valid separator string")
+		}
+		c.MetadataSeparatorList = getUniqueMetadataPatterns(c.MetadataSeparators)
+		sort.Stable(c.MetadataSeparatorList)
+	}
+	return nil
+}
```
How about fusing the two and reorganize the code a bit to save some indention level?
```suggestion
func (c *Config) initializeMetadataSeparators() error {
	// initialize metadata
	c.MetadataTags = map[string]string{}
	c.MetadataSeparatorList = []string{}

	if c.MetadataRows == 0 {
		return nil
	}
	if len(c.MetadataSeparators) == 0 {
		return fmt.Errorf("csv_metadata_separator required when specifying csv_metadata_rows")
	}

	patternList := map[string]bool{}
	for _, p := range patternMap {
		if patternList[p] {
			// Ignore further, duplicated entries
			continue
		}
		patternList[p] = true
		c.MetadataSeparatorList = append(c.MetadataSeparatorList, p)
	}
	sort.Stable(c.MetadataSeparatorList)
	return nil
}
```

srebhan commented on plugins/parsers/csv/parser.go:
```diff
@@ -17,33 +19,98 @@ import (
+	return nil
+}
+
+func (p *Parser) ParseMetadataRow(haystack string) map[string]string {
```
Please do not export this function.

srebhan commented on plugins/parsers/csv/parser.go:
```diff
@@ -67,6 +134,11 @@ func NewParser(c *Config) (*Parser, error) {
 		return nil, fmt.Errorf("csv_column_names field count doesn't match with csv_column_types")
 	}
 
+	err := c.initializeMetadataSeparators()
+	if err != nil {
+		return nil, err
+	}
```
```suggestion
	if err := c.initializeMetadataSeparators(); err != nil {
		return nil, fmt.Errorf("initializing separators failed: %v", err)
	}
```

srebhan commented on plugins/parsers/csv/parser.go:
```diff
@@ -117,15 +199,28 @@ func (p *Parser) ParseLine(line string) (telegraf.Metric, error) {
 	for p.SkipRows > 0 {
-		_, err := csvReader.Read()
-		if err != nil {
+		line, err := lineReader.ReadString('\n')
+		if err != nil && len(line) == 0 {
 			return nil, err
 		}
 		p.SkipRows--
```
I think this is wrong now as it does not decrement the `SkipRows` count if the line is empty.
```suggestion
		_, err := lineReader.ReadString('\n')
		if err != nil {
			return nil, err
		}
		p.SkipRows--
```

Hipska replied:
Could even be like this?
```suggestion
		if _, err := lineReader.ReadString('\n'); err != nil {
			return nil, err
		}
		p.SkipRows--
```

Hipska replied:
@etycomputer What do you think?

srebhan replied:
This code is still wrong. Imagine you have the following data
```csv
# My skip row stuff

yeah it even might contain totally empty lines




wooohhhaaa
time,value,tag
```
then all those empty lines (only containing `\n`) will return `nil, nil` in your code but **NOT** decrement `SkipRows`!

srebhan replied:
Even an empty line will contain the `\n` and thus has a length greater than zero. So the above code will work.

srebhan commented on plugins/parsers/csv/parser.go:
```diff
@@ -117,15 +199,28 @@ func (p *Parser) ParseLine(line string) (telegraf.Metric, error) {
 		p.SkipRows--
 	}
+	// Parse metadata
+	for p.MetadataRows > 0 {
+		line, err := lineReader.ReadString('\n')
+		if err != nil && len(line) == 0 {
+			return nil, err
+		}
+		p.MetadataRows--
```
Same as above
```suggestion
	for p.MetadataRows > 0 {
		line, err := lineReader.ReadString('\n')
		if err != nil {
			return nil, err
		}
		p.MetadataRows--
		if len(line) == 0 {
			continue
		}
```

srebhan reviewed (commented):
Thanks @etycomputer! I have some more comments, mostly about unexporting things. Please take a look.

srebhan commented on plugins/parsers/csv/parser.go:
```diff
@@ -69,6 +136,10 @@ func NewParser(c *Config) (*Parser, error) {
 	}
 
+	if err := c.initializeMetadataSeparators(); err != nil {
+		return nil, fmt.Errorf("Failed initializing separators. %v", err)
```
Error messages should start with a lowercase letter. How about
```suggestion
		return nil, fmt.Errorf("initializing separators failed: %v", err)
```

srebhan reviewed (commented):
@etycomputer thanks for the update. There are still some comments open from the last review and one more in this on. Can you please take a look?

Hipska commented on plugins/parsers/csv/parser.go:
```diff
@@ -46,6 +54,65 @@ type Parser struct {
 
+type MetadataPattern []string
+
+func (record MetadataPattern) Len() int {
+	return len(record)
+}
+func (record MetadataPattern) Swap(i, j int) {
+	record[i], record[j] = record[j], record[i]
+}
+func (record MetadataPattern) Less(i, j int) bool {
+	// Metadata with longer lengths should be ordered before shorter metadata
+	return len(record[i]) > len(record[j])
+}
```
I'm wondering if there are no already existing string array types that do provide this functionality.

Hipska replied:
@srebhan do you know this?

srebhan replied:
There is none that orders strings by their length AFAIK.

Hipska commented on plugins/parsers/csv/parser.go:
```diff
@@ -104,6 +175,16 @@ func (p *Parser) Parse(buf []byte) ([]telegraf.Metric, error) {
 // ParseLine does not use any information in header and assumes DataColumns is set
 // it will also not skip any rows
```
I think this comment is not correct anymore

Hipska commented:
It doesn't make much sense to ask for a review from us if we still have questions that are unanswered. The already done changes look good though. KUTGW!

srebhan reviewed (changes requested):
Hey @etycomputer, your `SkipRows` and `MetadataRows` handling is still not correct. I added an example where I think your code will fail... Please take a look!

etycomputer commented:
Hey @srebhan and @sspaink, I really appreciate the time you have put into this feature branch.
I had to rebase 3 times and every time there are a lot of merge conflicts. 
Could you please have a final look and if you are happy, accept the changes.

Kind regards,
Ehsan

etycomputer commented:
Hi @srebhan, I had a unit test that cover the Metadata example you mentioned but I have also updated the same unit test to also cover the empty row for skipped rows. Please look at the unit test ```TestParseCSVFileWithMetadata()```. The current code coves these scenarios at a higher layer.

Hipska commented on plugins/parsers/csv/parser_test.go:
```diff
@@ -827,3 +827,222 @@ corrupted_line
+		require.Equal(t, expectedTags[i], m.Tags())
+	}
+
+	p = &Parser{
+		HeaderRowCount:     1,
+		SkipRows:           2,
+		MetadataRows:       4,
+		Comment:            "#",
+		TagColumns:         []string{"type", "version"},
+		MetadataSeparators: []string{":", "="},
+		MetadataTrimSet:    " #",
+	}
+	err = p.Init()
+	require.NoError(t, err)
+	testCSVRows := []string{
+		"garbage nonsense that needs be skipped",
+		"",
+		"# version= 1.0\r\n",
+		"",
+		"    invalid meta data that can be ignored.\r\n",
+		"file created: 2021-10-08T12:34:18+10:00",
+		"timestamp,type,name,status\n",
+		"2020-11-23T08:19:27+10:00,Reader,R002,1\r\n",
+		"#2020-11-04T13:23:04+10:00,Reader,R031,0\n",
+		"2020-11-04T13:29:47+10:00,Coordinator,C001,0",
+	}
```
Why defining the same twice? Couldn't `p.Init()` be run a second time to reset internal counters? Also testCSVrows and the resulting tests could be reused imho.

srebhan replied:
No, `Init()` will not reset the counters because we are using the variable set by the config (i.e. the `MetadataRows` field) directly.

Hipska replied:
Okay, but still defining the test csv rows and the resulting tests twice seems unnecessary and error prone.

Hipska replied:
@etycomputer could you change to define the test csv only once please

Hipska replied:
Still not changed or given any comment on why not..

Hipska replied:
@srebhan I discussed with @etycomputer about this on Slack. Do you agree we should only test if the processor can handle csv's with metadata in this test only and maybe add another test that checks the compatibility of different line endings?

srebhan replied:
I'm fine either way.

srebhan commented:
@etycomputer I now see my fault. Even an empty line contains the `\n` which passes the `if` as the string is not empty. Sorry for the noise.

srebhan commented:
@etycomputer I'm hoping to fix your CI error with #10497. After this (and a rebase from your side) I guess we are good to go.

etycomputer commented:
> @etycomputer I now see my fault. Even an empty line contains the `\n` which passes the `if` as the string is not empty. Sorry for the noise.

No problem at all. It's your job to ask and mine to make sure everything is answered.

etycomputer commented:
Hi @srebhan and @Hipska,

I have rebased my branch and fixed all CI issues.

It is ready for you to do the final review.

I did have to fix a test bug introduced on 04/02/2022 by Rabhan. 

```
const simpleCSVWithHeader = `
# Simple CSV with header(s)
a,b,c
1.2,3.1415,ok
`
```

```
parser := &csv.Parser{
	MetricName:  "metricName",
	SkipRows:    2,
	ColumnNames: []string{"a", "b", "c"},
	TagColumns:  []string{"c"},
}
```
The issue is that there is an empty line at the start and one comment followed by a header and he has only two SkipRows and no header rows.

I have fixed the test by changing the SkipRows to 3

srebhan reviewed (approved):
Looks good to me. Thanks for working on this @etycomputer!

Hipska commented on plugins/inputs/http/http_test.go:
```diff
@@ -377,7 +377,7 @@ func TestHTTPWithCSVFormat(t *testing.T) {
 		parser := &csv.Parser{
 			MetricName:  "metricName",
-			SkipRows:    2,
+			SkipRows:    3,
```
Does this mean people using `csv_skip_rows` will need to change their config?

sspaink reviewed (changes requested):
Thank you for working on this, I have two questions for you if you could help answer them.

srebhan commented on plugins/parsers/csv/parser.go:
```diff
@@ -99,9 +168,17 @@ func (p *Parser) Parse(buf []byte) ([]telegraf.Metric, error) {
-// it will also not skip any rows
 func (p *Parser) ParseLine(line string) (telegraf.Metric, error) {
+	if len(line) == 0 {
+		if p.SkipRows > 0 {
```
Maybe I can explain it @sspaink... We account for the case where you have a file and call `ParseLine` multiple times. For example, imaging a plugin (e.g. tail or file) to read the file line-by-line and call `ParseLine` each time a new line arrives. In this case you do have a state. You first should do nothing (i.e. skip) for the first `SkipRows` number of lines, later maybe read `N` header lines etc. Without reflecting this state in the parser it is hard to keep track of where you are (skipping, reading metadata or headers or data) in the process. Actually this is nothing invented by this PR, but it is already in the code.

We can now argue if it is a good idea to overwrite the user-specified config or if it would have been better to add another fields, but that's not the fault of this PR...

sspaink replied:
I appreciate the explanation, I was confused about how SkipRows worked.

Hipska replied:
That's a good point @srebhan:
>We can now argue if it is a good idea to overwrite the user-specified config or if it would have been better to add another fields, but that's not the fault of this PR...

What happens if telegraf config gets reloaded while still running? It will skip x rows again in that case?

srebhan commented on plugins/parsers/csv/parser.go:
```diff
@@ -99,9 +168,17 @@ func (p *Parser) Parse(buf []byte) ([]telegraf.Metric, error) {
+	if len(line) == 0 {
+		if p.SkipRows > 0 {
+			p.SkipRows--
+			return nil, io.EOF
```
Same as above. You currently do have
```golang
	for p.SkipRows > 0 {
		_, err := csvReader.Read()
		if err != nil {
			return nil, err
		}
		p.SkipRows--
	}
```
in `parseCSV` line 122ff. The `Read()` function will return `io.EOF` if there is no more line to read but we are expecting one... The code above just replicates this behavior.

sspaink replied:
Thanks for pointing that out, makes sense

etycomputer commented:
Thanks, everyone; it has been a long journey.
I appreciate your commitment and contributions.

srebhan commented:
@etycomputer and I appreciate your commitment and hard work to get this over the finishing line!

JackSharebourg commented:
Hey @etycomputer thanks for implementing this feat!

In the merged version you put all metadata in tags, but the comment and the Readme still point out that metadata will be fields by default. 
IMHO I would also recommend to change this back. Metadata should be fields by default because this is also the current way for all columns to be parsed and with the tag key list they can easily be converted when needed. 
This way metadata could also be used as the measurement, since the measurement is chosen only out of fields and not out of tags.
Using a processor to change metadata tags back to fields or as measurement every time this is needed is much more complicated and error prone then putting the metadata tags in the tag list. @Hipska what do you think?

Hipska commented:
The readme should definitely be updated to reflect the actual scenario! Please create a new issue for that.

This was my opinion back when we started reviewing this PR, it has not been changed since then:
> General idea looks good. Only it looks more logical that metadata would end up in metric _tags_ instead of metric _fields_. Or maybe have it as an option if you don't agree.
